proposals/NNNN-async-stream-continuation-hashable-conformance.md

```diff
@@ -0,0 +1,138 @@
+# Feature name
+
+* Proposal: [SE-NNNN](NNNN-async-stream-continuation-hashable-conformance.md)
+* Authors: [Mykola Pokhylets](https://github.com/nickolas-pohilets)
+* Review Manager: TBD
+* Status: **Awaiting review**
+* Implementation: [swiftlang/swift#79457](https://github.com/swiftlang/swift/pull/79457)
+* Review: ([pitch](https://forums.swift.org/t/pitch-add-hashable-conformance-to-asyncstream-continuation/77897))
+
+## Introduction
+
+This proposal suggests to add `Hashable` conformance to `Async(Throwing)Stream.Continuation`
+to simplify working with multiple streams.
+
+## Motivation
+
+Use cases operating with multiple `AsyncStream`'s may need to store multiple continuations.
+When handling `onTermination` callback, client code needs to remove relevant continuation.
+
+To identify relevant continuation, client code needs to be able to compare continuations.
+
+It is possible to associate lookup key with each continuation, but this is inefficient.
+`AsyncStream.Continuation` already stores a reference to `AsyncStream._Storage`,
+whose identity can be used to provide simple and efficient `Hashable` conformance.
+
+Consider this simple Observer pattern with `AsyncSequence`-based API.
+To avoid implementing `AsyncSequence` from scratch it uses `AsyncStream` as a building block.
+To support multiple subscribes new stream is returned every time.
+
+```swift
+@MainActor private class Sender {
+    var value: Int = 0 {
+        didSet {
+            for c in continuations {
+                c.yield(value)
+            }
+        }
+    }
+
+    var values: some AsyncSequence<Int, Never> {
+        AsyncStream<Int>(bufferingPolicy: .bufferingNewest(1)) { continuation in
+            continuation.yield(value)
+            self.continuations.insert(continuation)
+            continuation.onTermination = { _ in
+                DispatchQueue.main.async {
+                    self.continuations.remove(continuation)
+                }
+            }
+        }
+    }
+
+    private var continuations: Set<AsyncStream<Int>.Continuation> = []
+}
+```
+
+Without `Hashable` conformance, continuation needs to be associated with an artificial identifier,
+e.g. object identity by wrapping each continuation in a class:
```

Comment:
```suggestion
(e.g., object identity) by wrapping each continuation in a class:
```

Reply:
"wrapping" is still part of the example. Wrapping in a class is not required, there are alternatives, e.g. one could store tuples of `(continuation, UUID)`.